### DISCUSSION TITLE
Archiecture design

### DISCUSSION THREAD

**Discussion by tologonkudaiberdiuulu** [robertodoering/harpy#650] [General]
Hi, @robertodoering . I have a question related to service locator. Why you've removed get_it from project? Is something wrong with get_it? Or it's bad dependency injection with it? Can you explain your decision?. Thank you in advance

**Comment by robertodoering** [comment]
Hey @tologonkudaiberdiuulu,
I recently introduced riverpod to the project and at the same time removed `provider`, `flutter_bloc` and `get_it`.
get_it is fine as a simple service locator but there is no reason to use it in combination with riverpod as you can provide and mock services with it as well.

Additionally, riverpod is much more powerful, especially when services depend on other services.
For example in harpy we have a `twitterApiProvider` which needs to instantiate a `TwitterApi` based on the `userToken`. 
Whenever the token changes (e.g. when logging out or changing user) we need to use a new `TwitterApi` instance with the updated token.

With riverpod we can just watch the `authPreferencesProvider` to get the `authToken`. Now whenever the `authPreferencesProvider` provides a new `authToken` the `twitterApiProvider` will automatically instantiate a new `TwitterApi` with the updated token.

Previously we had to manually update the TwitterApi to update the `userToken`.

**Comment by tologonkudaiberdiuulu** [comment]
Ok. Got it. I noticed that some people(youtuber or blogger, I don't remeember) hating get_it. I thought you also didn't like get_it. And one more question, what do you recommended for big projects(production) as a service locator? Or do you have any clean architecture tips?

**Reply by robertodoering** [reply]
Generally I really like riverpod and would go with that to provide my services as I do in harpy.
